scylla/src/transport/connection.rs

```diff
@@ -1664,4 +1680,89 @@ mod tests {
             Err(QueryError::BadQuery(BadQuery::Other(_)))
         ));
     }
+
+    #[tokio::test]
+    async fn test_lwt_optimisation_mark_negotiation() {
+        const MASK: &str = "2137";
+
+        fn lwt_optimisation_entry() -> String {
+            format!("{}={}", LWT_OPTIMIZATION_META_BIT_MASK_KEY, MASK)
+        }
+
+        let proxy_addr = SocketAddr::from_str("127.0.0.54:9042").unwrap();
+
+        let config = ConnectionConfig::default();
+
+        let (startup_tx, mut startup_rx) = mpsc::unbounded_channel();
+
+        let options_without_lwt_optimisation_support =
+            [].into_iter().collect::<HashMap<String, Vec<String>>>();
+        let options_with_lwt_optimisation_support = [(
+            SCYLLA_LWT_ADD_METADATA_MARK_EXTENSION.into(),
+            vec![lwt_optimisation_entry()],
+        )]
+        .into_iter()
+        .collect::<HashMap<String, Vec<String>>>();
+
+        let make_rules = |options| {
+            vec![
+                RequestRule(
+                    Condition::RequestOpcode(RequestOpcode::Options),
+                    RequestReaction::forge_response(Arc::new(move |frame: RequestFrame| {
+                        ResponseFrame::forged_supported(frame.params, &options).unwrap()
+                    })),
+                ),
+                RequestRule(
+                    Condition::RequestOpcode(RequestOpcode::Startup),
+                    RequestReaction::drop_frame().with_feedback_when_performed(startup_tx.clone()),
+                ),
+            ]
+        };
+
+        let mut proxy = Proxy::builder()
+            .with_node(
+                Node::builder()
+                    .proxy_address(proxy_addr)
+                    .request_rules(make_rules(options_without_lwt_optimisation_support))
+                    .build_dry_mode(),
+            )
+            .build()
+            .run()
+            .await
+            .unwrap();
+
+        // We must interrupt the driver's full connection opening, because our proxy does not interact further after Startup.
```

Comment:
Why do we need to interrupt it? What happens if the connection opening process proceeds too far before the timeout? What guarantee do we have that 10ms isn't too short or too long?

Couldn't we intercept the `Options` and `Startup` messages and modify the LWT bit mask there, then let the real Scylla handle the rest? The LWT mask doesn't introduce large changes to the protocol - unless you send LWT queries explicitly everything should be fine.

Reply:
> Why do we need to interrupt it? What happens if the connection opening process proceeds too far before the timeout? What guarantee do we have that 10ms isn't too short or too long?

You are right here, we have no such guarantee. I'll fix this with a `select!` over `open_connection` and `startup_rx`.

> Couldn't we intercept the `Options` and `Startup` messages and modify the LWT bit mask there, then let the real Scylla handle the rest? The LWT mask doesn't introduce large changes to the protocol - unless you send LWT queries explicitly everything should be fine.

I believe the purpose of such unit tests is to minimize amount of actions performed and isolate the specific part; in this case, further messages between driver and cluster are needless for this test's aim.

Reply:
> You are right here, we have no such guarantee. I'll fix this with a `select!` over `open_connection` and `startup_rx`.

You should also make sure that the proxy doesn't respond to any requests after the `Startup` message. I'm not sure how the proxy reacts to requests when it is in dry mode and none of the rules match, but just ignoring them (not returning any response) should work here.

Reply:
That's exactly the default behaviour of a proxy in dry mode - ignoring requests.